### ISSUE TITLE
Connection failures following control plane upgrade to 1.6

### ISSUE THREAD

**Issue by Stono** [istio/istio#28026]
**Bug description**
Another set of apps that doesn't work.  We upgrade the control plane to `1.6`, but don't touch the apps (so still running `1.5` sidecar) and we start getting the old `upstream connect error or disconnect/reset before headers. reset reason: connection failure` when making requests from within the mesh (from another app to this one).

Looking at the istio-proxy debug logs for the destination service we see:
```
[Envoy (Epoch 0)] [2020-10-16 07:46:36.184][31][debug][filter] [external/envoy/source/extensions/filters/listener/original_dst/original_dst.cc:18] original_dst: New connection accepted
[Envoy (Epoch 0)] [2020-10-16 07:46:36.184][31][debug][conn_handler] [external/envoy/source/server/connection_handler_impl.cc:361] [C83] new connection
[Envoy (Epoch 0)] [2020-10-16 07:46:36.184][31][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:279] [C83] dispatch error: http/1.1 protocol error: HPE_INVALID_METHOD
[Envoy (Epoch 0)] [2020-10-16 07:46:36.184][31][debug][connection] [external/envoy/source/common/network/connection_impl.cc:101] [C83] closing data_to_write=66 type=2
[Envoy (Epoch 0)] [2020-10-16 07:46:36.184][31][debug][connection] [external/envoy/source/common/network/connection_impl_base.cc:30] [C83] setting delayed close timer with timeout 1000 ms
[Envoy (Epoch 0)] [2020-10-16 07:46:36.184][31][debug][connection] [external/envoy/source/common/network/connection_impl.cc:600] [C83] write flush complete
```

Here are config dumps from `istio-proxy` on the destination service from 1.5 and 1.6 control planes:
[config-dumps.tar.gz](https://github.com/istio/istio/files/5390054/config-dumps.tar.gz)

The pod exposes three ports:
```
      ports:
      - containerPort: 9080
        name: http-web-admin
        protocol: TCP
      - containerPort: 8317
        name: http-web
        protocol: TCP
      - containerPort: 8318
        name: tcp-tcp-sftp
        protocol: TCP
```

```
❯ k -n vds-acceptor get ep
NAME    ENDPOINTS           AGE
admin   10.198.10.11:9080   338d
app     10.198.10.11:8317   435d
sftp    10.198.10.11:8318   356d
```

The cluster is configured with strict mtls:
```
apiVersion: security.istio.io/v1beta1
kind: PeerAuthentication
metadata:
  name: default
  namespace: istio-system
spec:
  mtls:
    mode: STRICT
```

The only difference I could draw was that this app (and the other broken services) also has a non-mesh service (`stfp`):

However that is excluded from istio with a `Sidecar`:
```
  ingress:
  - defaultEndpoint: 127.0.0.1:8317
    port:
      name: http-app
      number: 8317
      protocol: HTTP
  - defaultEndpoint: 127.0.0.1:9080
    port:
      name: http-admin
      number: 9080
      protocol: HTTP
```

and
```
traffic.sidecar.istio.io/includeInboundPorts: 8317,9080
```

We tried adding an explicit `PeerAuthentication` policy (even though it shouldn't actually make any difference):
```
apiVersion: security.istio.io/v1beta1
kind: PeerAuthentication
metadata:
  name: default
spec:
  mtls:
    mode: STRICT
  portLevelMtls:
    8318:
      mode: DISABLE
  selector:
    matchLabels:
      app: vds-acceptor
      group: delivery-platform
```

And as expected it didn't change anything.

**Restarting the app to pull in a 1.6 sidecar fixed the issue**, therefore it seems to only be an issue with a 1.5 sidecar & 1.6 control plane.

[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[x] Test and Release
[x] User Experience
[ ] Developer Infrastructure

**Expected behavior**
Upgrades to not break services

**Steps to reproduce the bug**
🤷  happy to live debug my cluster with you

**Version (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version` if you used Helm)**
1.6.11

**How was Istio installed?**
helm

**Environment where bug was observed (cloud vendor, OS, etc)**
gke

**Comment by Stono** [comment]
cc. @howardjohn @mandarjog @rshriram any help would be appreciated.  Happy to give you access to the test cluster (or screen share if it helps).

**Comment by mandarjog** [comment]
A couple of things I can think of are h2upgrade differences or TLS mismatch.

@PiotrSikora hpe_invalid_method indicates that it chose the http1 codec.

**Comment by Stono** [comment]
@mandarjog : no http2, we still have it disabled following the last incident.

**Comment by Stono** [comment]
@mandarjog @howardjohn sorry to keep tagging you but i'm really keen to get rid of this 1.6 migration.  This is the last remaining "blocking" issue for us as it causes a service outage until the pods are restarted.

I can recreate it easily, so will work with anyone to get to the bottom of it.

**Comment by Stono** [comment]
OK we can confirm any workload (regardless of it its port combination) using `traffic.sidecar.istio.io/includeInboundPorts: "8080"` breaks during a 1.5 to 1.6 upgrade.

So to recreate this problem:

1. Deploy 1.5 cluster
2. Deploy a workload with a single port (8080) with MTLS enabled, with the annotation `traffic.sidecar.istio.io/includeInboundPorts: "8080"`
3. Upgrade control plane to 1.6

 = broken

4. Restart workload

 = fixed

**Comment by Stono** [comment]
So we continued to try and narrow this down.  We did the following:

1. Deploy a 1.5 cluster
2. Deploy a 1.5 workload with mTLS enabled, single port (8080) with `traffic.sidecar.istio.io/includeInboundPorts: 8080`
3. Upgrade control plane to 1.6

 = Broken

The errors were slightly different on this one:
```
[Envoy (Epoch 0)] [2020-10-19 12:59:38.818][28][debug][filter] [external/envoy/source/extensions/filters/listener/original_dst/original_dst.cc:18] original_dst: New connection accepted
[Envoy (Epoch 0)] [2020-10-19 12:59:38.818][28][debug][filter] [external/envoy/source/common/tcp_proxy/tcp_proxy.cc:232] [C1527] new tcp proxy session
[Envoy (Epoch 0)] [2020-10-19 12:59:38.818][28][debug][filter] [external/envoy/source/common/tcp_proxy/tcp_proxy.cc:381] [C1527] Creating connection to cluster BlackHoleCluster
[Envoy (Epoch 0)] [2020-10-19 12:59:38.818][28][debug][upstream] [external/envoy/source/common/upstream/cluster_manager_impl.cc:1283] no healthy host for TCP connection pool
[Envoy (Epoch 0)] [2020-10-19 12:59:38.818][28][debug][connection] [external/envoy/source/common/network/connection_impl.cc:101] [C1527] closing data_to_write=0 type=1
[Envoy (Epoch 0)] [2020-10-19 12:59:38.818][28][debug][connection] [external/envoy/source/common/network/connection_impl.cc:192] [C1527] closing socket: 1
[Envoy (Epoch 0)] [2020-10-19 12:59:38.821][28][debug][filter] [external/envoy/source/extensions/filters/listener/original_dst/original_dst.cc:18] original_dst: New connection accepted
[Envoy (Epoch 0)] [2020-10-19 12:59:38.822][28][debug][filter] [external/envoy/source/common/tcp_proxy/tcp_proxy.cc:232] [C1528] new tcp proxy session
[Envoy (Epoch 0)] [2020-10-19 12:59:38.822][28][debug][filter] [external/envoy/source/common/tcp_proxy/tcp_proxy.cc:381] [C1528] Creating connection to cluster BlackHoleCluster
[Envoy (Epoch 0)] [2020-10-19 12:59:38.822][28][debug][upstream] [external/envoy/source/common/upstream/cluster_manager_impl.cc:1283] no healthy host for TCP connection pool
[Envoy (Epoch 0)] [2020-10-19 12:59:38.822][28][debug][connection] [external/envoy/source/common/network/connection_impl.cc:101] [C1528] closing data_to_write=0 type=1
[Envoy (Epoch 0)] [2020-10-19 12:59:38.822][28][debug][connection] [external/envoy/source/common/network/connection_impl.cc:192] [C1528] closing socket: 1
[Envoy (Epoch 0)] [2020-10-19 12:59:38.873][28][debug][filter] [external/envoy/source/extensions/filters/listener/original_dst/original_dst.cc:18] original_dst: New connection accepted
[Envoy (Epoch 0)] [2020-10-19 12:59:38.873][28][debug][filter] [external/envoy/source/common/tcp_proxy/tcp_proxy.cc:232] [C1529] new tcp proxy session
[Envoy (Epoch 0)] [2020-10-19 12:59:38.873][28][debug][filter] [external/envoy/source/common/tcp_proxy/tcp_proxy.cc:381] [C1529] Creating connection to cluster BlackHoleCluster
[Envoy (Epoch 0)] [2020-10-19 12:59:38.873][28][debug][upstream] [external/envoy/source/common/upstream/cluster_manager_impl.cc:1283] no healthy host for TCP connection pool
[Envoy (Epoch 0)] [2020-10-19 12:59:38.873][28][debug][connection] [external/envoy/source/common/network/connection_impl.cc:101] [C1529] closing data_to_write=0 type=1
[Envoy (Epoch 0)] [2020-10-19 12:59:38.873][28][debug][connection] [external/envoy/source/common/network/connection_impl.cc:192] [C1529] closing socket: 1
[Envoy (Epoch 0)] [2020-10-19 12:59:38.907][28][debug][filter] [external/envoy/source/extensions/filters/listener/original_dst/original_dst.cc:18] original_dst: New connection accepted
[Envoy (Epoch 0)] [2020-10-19 12:59:38.907][28][debug][filter] [external/envoy/source/common/tcp_proxy/tcp_proxy.cc:232] [C1530] new tcp proxy session
[Envoy (Epoch 0)] [2020-10-19 12:59:38.907][28][debug][filter] [external/envoy/source/common/tcp_proxy/tcp_proxy.cc:381] [C1530] Creating connection to cluster BlackHoleCluster
[Envoy (Epoch 0)] [2020-10-19 12:59:38.907][28][debug][upstream] [external/envoy/source/common/upstream/cluster_manager_impl.cc:1283] no healthy host for TCP connection pool
[Envoy (Epoch 0)] [2020-10-19 12:59:38.907][28][debug][connection] [external/envoy/source/common/network/connection_impl.cc:101] [C1530] closing data_to_write=0 type=1
[Envoy (Epoch 0)] [2020-10-19 12:59:38.907][28][debug][connection] [external/envoy/source/common/network/connection_impl.cc:192] [C1530] closing socket: 1
[Envoy (Epoch 0)] [2020-10-19 12:59:40.096][20][debug][main] [external/envoy/source/server/server.cc:177] flushing stats
[Envoy (Epoch 0)] [2020-10-19 12:59:42.677][20][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:263] [C0] new stream
[Envoy (Epoch 0)] [2020-10-19 12:59:42.677][20][debug][http] [external/envoy/source/common/http/conn_manager_impl.cc:732] [C0][S5899294795249759883] request headers complete (end_stream=true):
```

**Comment by Stono** [comment]
To update further, we tried again with two workloads:

- workload1: port 8080, `traffic.sidecar.istio.io/includeInboundPorts: 8080`
- workload2: port 8080, no annotation

Updated control plane to 1.6

- workload1: broken
- workload2: working

Restart workload 2 = fixed.

Therefore i'm 100% confident the annotation is the cause during the upgrade process.

The reason this is so impacting for us it we have it on _every_ workload.

**Comment by howardjohn** [comment]
Good pod:
```
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15006
```
```yaml
      - istio-iptables
      - -p
      - "15001"
      - -z
      - "15006"
      - -u
      - "1337"
      - -m
      - REDIRECT
      - -i
      - ...
      - -x
      - ...
      - -b
      - -d
      - "15020"
      - -o
      - ...
```

Bad pod:
```
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15001
```
```yaml
    - command:
      - istio-iptables
      - -p
      - "15001"
      - -z
      - "15006"
      - -u
      - "1337"
      - -m
      - REDIRECT
      - -i
      - ....
      - -x
      - foo,bar
      - -b
      - "8080"
      - -d
      - "15020"
      - -o
      - x,y,z,not-relevant
```

The interesting thing is the arguments actually show `-z 15006` - but not used. I think this was some migration thing that is going wrong here

**Comment by howardjohn** [comment]
On 1.5:
```
$ istio-iptables -n -p 15001 -z 15006 -b 8080
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15001
```

On 1.6:
```
$ istio-iptables -n -p 15001 -z 15006 -b 8080
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15001
```

**Comment by howardjohn** [comment]
Missing #22686?

**Comment by howardjohn** [comment]
Works in 1.5 pilot because https://github.com/istio/istio/commit/ab88b62363307f92f88d48e23a4fa2e14a3df88d is not in 1.5

**Comment by howardjohn** [comment]
I am not sure I am seeing a workaround for this (aside from removing the annotation) other than cherrypicking https://github.com/istio/istio/pull/22686  or reverting ab88b62  unfortunately... There is just a mismatch between these two, the latter assumes the former was always present when its not during the upgrade case

**Comment by Stono** [comment]
I've tested @howardjohn fix in #28111 and can confirm it works as expected

**Comment by howardjohn** [comment]
Thanks @Stono ! For anyone else that runs into this, please use the latest 1.6 and PILOT_ENABLE_LEGACY_INBOUND_LISTENERS, or 1.7+ is fixed natively. Thanks!
